src/icp_v2.cc

```diff
@@ -669,39 +655,56 @@ icpHandleUdp(int sock, void *)
-
+                throw TextException(xstrerr(xerrno), Here());
             break;
         }
 
         ++(*N);
         icpCount(buf, RECV, (size_t) len, 0);
         buf[len] = '\0';
-        debugs(12, 4, "icpHandleUdp: FD " << sock << ": received " <<
-               (unsigned long int)len << " bytes from " << from);
-
-#ifdef ICP_PACKET_DUMP
 
-        icpPktDump(buf);
-#endif
-
-        if ((size_t) len < sizeof(icp_common_t)) {
-            debugs(12, 4, "icpHandleUdp: Ignoring too-small UDP packet");
-            break;
-        }
-
-        icp_version = (int) buf[1]; /* cheat! */
+        // XXX: possible unaligned integer access
+        auto pkt = reinterpret_cast<icp_common_t *>(buf);
+        auto payloadLength = ntohs(pkt->length);
+
+        if (static_cast<size_t>(len) < sizeof(icp_common_t))
+            throw TextException("too-small UDP packet (ignoring)", Here());
+
+        if (payloadLength < len-sizeof(icp_common_t))
+            throw TextException("length value larger than data received (ignoring)", Here());
+
+        // TODO use BinaryTokenizer parse to validate fully
+        if (pkt->opcode < 0 || pkt->opcode >= ICP_END)
+            throw TextException(ToSBuf("unknown opcode ", pkt->opcode), Here());
+
+        struct in_addr a;
+        a.s_addr = pkt->shostid;
+        debugs(12, 2, "ICP Client remote=" << from << " FD " << sock);
+        debugs(12, 2, "ICP Client REQUEST:\n---------\n" <<
+               "opcode=(" << std::setw(3) << pkt->opcode << ")" << icp_opcode_str[pkt->opcode] <<
+               ", version=" << pkt->version <<
+               ", length=" << payloadLength <<
+               ", reqnum=" << ntohl(pkt->reqnum) <<
+               ", flags=" << std::hex << ntohl(pkt->flags) << std::dec <<
+               ", shostid=" << Ip::Address(a) <<
+               Debug::Extra <<
+               Raw(nullptr, (buf+sizeof(icp_common_t)), payloadLength).minLevel(DBG_DATA).hex() <<
+               "\n----------");
 
         if (icpOutgoingConn->local == from)
             // ignore ICP packets which loop back (multicast usually)
-            debugs(12, 4, "icpHandleUdp: Ignoring UDP packet sent by myself");
-        else if (icp_version == ICP_VERSION_2)
+            throw TextException("UDP packet sent by myself", Here());
+        else if (pkt->version == ICP_VERSION_2)
             icpHandleIcpV2(sock, from, buf, len);
-        else if (icp_version == ICP_VERSION_3)
+        else if (pkt->version == ICP_VERSION_3)
             icpHandleIcpV3(sock, from, buf, len);
         else
-            debugs(12, DBG_IMPORTANT, "WARNING: Unused ICP version " << icp_version <<
-                   " received from " << from);
+            throw TextException(ToSBuf("unknown version ", pkt->version), Here());
+
+        } catch (...) {
+            debugs(12, 2, "WARNING: ICP remote=" << from << " FD " << sock << ": " << CurrentException);
```

Comment:
This hides all errors from the admin, including previously reported "abnormal" or "unexpected" errors from trusted peers. Even ICP access logs will not disclose them AFAICT. Please update the PR description to disclose and justify this significant change.

Reply:
Reverted the in meaningful debugs error message in 36d79a35f.

The high level message about unknown ICP version being ignored does not need to be logged at level-1.